Report/sections/3_implementation/astbuilder.tex

```diff
@@ -0,0 +1,113 @@
+Note that a list structure could have been used to store the AST's, but we chose dictionaries for the faster lookup.
+
+Once all the individual AST's have been produced, they are combined by giving them a common root node which can be seen on line \ref{line:newRoot} and a combined AST, with the new root, is created on line \ref{line:CombinedAST}.
+ 
+\subsubsection*{Visiting the Game Object}
+In the following, we will further describe how the \textbf{VisitGameObject} method from \snipref{lst:BuildAstMethod} on line \ref{line:VisitGO} works and how the over all visitor pattern comes into play.
+
+As \textbf{VisitGameObject} gets called, with the parse tree in its parameter, the initial step is to perform pattern matching on the type of the \textbf{gameObject} which can be seen in the definition for \texttt{VisitGameObject} on line \ref{line:PatternMatch} in \snipref{lst:VisitParseTree}. This is done to ensure that the parsed source code contains a game object of the correct type and if that is the case, create a node of either of the legal types which is then assigned to the local variable \textbf{typeNode}.
+
+Once the pattern matching is done we use object initialization to create a new \textbf{gameObjectNode} node on line \ref{line:NewGONode}. In the initialization of this object we call \texttt{VisitGameObjectContents} method on line \ref{line:VisitGameObjectContents} which is where the cascading effect of the visitor pattern begins.
+
+\begin{lstlisting}[caption={The initial stage of visiting}, label={lst:VisitParseTree},escapechar=|]
+    public GameObjectNode VisitGameObject(DazelParser.GameObjectContext context)
+    {
+        GameObjectTypeNode typeNode;
+        
+        switch (context.gameObjectType.Type) |\label{line:PatternMatch}|
+        {
+            case DazelLexer.SCREEN:
+                typeNode = new ScreenNode();
+                break;
+            case DazelLexer.ENTITY:
+                typeNode = new EntityNode();
+                break;
+            case DazelLexer.MOVE_PATTERN:
+                typeNode = new MovePatternNode();
+                break;
+            default:
+                throw new ArgumentException("Type is not a GameObjectType!");
+        }
+
+        GameObjectNode gameObjectNode = new GameObjectNode() |\label{line:NewGONode}|
+        {
+            Identifier = context.GetChild(1).GetText(),
+            TypeNode = typeNode,
+            Contents = VisitGameObjectContents(context.gameObjectBlock().gameObjectContents()) |\label{line:VisitGameObjectContents}|
+        };
+        
+        return gameObjectNode;
+    }
+\end{lstlisting}
+
+
+Durings visitation of the parse tree the visitor will reduce the parse tree to an AST. It does this by skipping all none terminal nodes and only constructing the AST of terminal nodes. An example of this can be seen in \snipref{lst:VisitAssignment}. 
```

Comment:
```suggestion
Durings visitation of the parse tree the visitor will reduce the parse tree to an AST. It does this by skipping all nonterminal nodes and only constructing the AST of terminal nodes. An example of this can be seen in \snipref{lst:VisitAssignment}. 
```